app/models/daos/OpenIDInfoDAO.scala

```diff
@@ -0,0 +1,117 @@
+package models.daos
+
+import com.mohiva.play.silhouette.api.LoginInfo
+import com.mohiva.play.silhouette.impl.daos.DelegableAuthInfoDAO
+import com.mohiva.play.silhouette.impl.providers.OpenIDInfo
+import javax.inject.Inject
+import play.api.libs.concurrent.Execution.Implicits._
+import play.api.db.slick.DatabaseConfigProvider
+import scala.concurrent.Future
+
+/**
+ * The DAO to store the OpenID information.
+ */
+class OpenIDInfoDAO @Inject() (protected val dbConfigProvider: DatabaseConfigProvider)
+    extends DelegableAuthInfoDAO[OpenIDInfo] with DAOSlick {
+
+  import driver.api._
+
+  protected def openIDInfoQuery(loginInfo: LoginInfo) = for {
+    dbLoginInfo <- loginInfoQuery(loginInfo)
+    dbOpenIDInfo <- slickOpenIDInfos if dbOpenIDInfo.loginInfoId === dbLoginInfo.id
+  } yield dbOpenIDInfo
+
+  protected def addAction(loginInfo: LoginInfo, authInfo: OpenIDInfo) =
+    loginInfoQuery(loginInfo).result.head.flatMap { dbLoginInfo =>
+      DBIO.seq(
+        slickOpenIDInfos += DBOpenIDInfo(authInfo.id, dbLoginInfo.id.get),
+        slickOpenIDAttributes ++= authInfo.attributes.map {
+          case (key, value) => DBOpenIDAttribute(authInfo.id, key, value)
+        })
+    }.transactionally
+
+  protected def updateAction(loginInfo: LoginInfo, authInfo: OpenIDInfo) =
+    openIDInfoQuery(loginInfo).result.head.flatMap { dbOpenIDInfo =>
+      DBIO.seq(
+        slickOpenIDInfos filter(_.id === dbOpenIDInfo.id) update dbOpenIDInfo.copy(id = authInfo.id),
+        slickOpenIDAttributes.filter(_.id === dbOpenIDInfo.id).delete,
+        slickOpenIDAttributes ++= authInfo.attributes.map {
+          case (key, value) => DBOpenIDAttribute(authInfo.id, key, value)
+        })
+    }.transactionally
+
+  /**
+   * Finds the auth info which is linked with the specified login info.
+   *
+   * @param loginInfo The linked login info.
+   * @return The retrieved auth info or None if no auth info could be retrieved for the given login info.
+   */
+  def find(loginInfo: LoginInfo): Future[Option[OpenIDInfo]] = {
+    val query = openIDInfoQuery(loginInfo).joinLeft(slickOpenIDAttributes).on(_.id === _.id)
+    val result = db.run(query.result)
+    result.map { openIDInfos =>
+      if (openIDInfos.isEmpty) None
+      else {
+        val attrs = openIDInfos.collect { case (id, Some(attr)) => (attr.key, attr.value) }.toMap
+        Some(OpenIDInfo(openIDInfos.head._1.id, attrs))
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Usage of head on collections.](https://www.codacy.com/app/HiP-App/HiPCMS/file/2281812354/issues/source?bid=3144114&fileBranchId=3144125#l56)